include/NeoFOAM/finiteVolume/cellCentred/boundary/volumeBoundaryBase.hpp

```diff
@@ -4,25 +4,86 @@
 
 #include "NeoFOAM/fields/fieldTypeDefs.hpp"
 #include "NeoFOAM/fields/boundaryFields.hpp"
+#include "NeoFOAM/finiteVolume/cellCentred/boundary/boundaryBase.hpp"
+#include "NeoFOAM/core/registerClass.hpp"
 #include "NeoFOAM/mesh/unstructured/unstructuredMesh.hpp"
+#include "NeoFOAM/core/dictionary.hpp"
 
 namespace NeoFOAM::finiteVolume::cellCentred
 {
+// forward declaration so we can use it to define the create function and the class manager
+template<typename ValueType>
+class VolumeBoundaryModel;
+
+// define the create function use to instantiate the derived classes
+template<typename ValueType>
+using createFunc = std::function<std::unique_ptr<VolumeBoundaryModel<ValueType>>(
+    const UnstructuredMesh& mesh, const Dictionary dict, int patchID
+)>;
+
+// define the class manager to register the classes
+template<typename ValueType>
+using VolumeBoundaryModelManager =
+    NeoFOAM::BaseClassRegistry<VolumeBoundaryModel<ValueType>, createFunc<ValueType>>;
+
+template<typename ValueType>
+class VolumeBoundaryModel : public VolumeBoundaryModelManager<ValueType>
+{
+public:
+
+    template<typename derivedClass>
+    using VolumeBoundaryModelReg = NeoFOAM::
+        RegisteredClass<derivedClass, VolumeBoundaryModel<ValueType>, createFunc<ValueType>>;
+
+    template<typename derivedClass>
+    bool registerClass()
```

Comment:
if not used in static context i think this can be marked const?
```suggestion
    bool registerClass() const
```